src/config.rs

```diff
@@ -0,0 +1,405 @@
+    ///
+    /// let config = Config::from(Env::prefixed("some_prefix")).unwrap();
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// Any [Error] occurring while extracting the configuration will be
+    /// returned.
+    pub fn from<T: Provider>(provider: T) -> Result<Config, Error> {
+        Figment::from(provider).extract()
+    }
+
+    /// Provide a default provider, a `Figment`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use libpacstall::config::Config;
+    ///
+    /// let figment = Config::figment();
+    /// ```
+    pub fn figment() -> Figment {
+        Figment::from(Toml::file("/etc/pacstall/config.toml"))
+            .merge(Env::prefixed("PACSTALL_").split("_"))
+    }
+}
+
+impl figment::Provider for Config {
+    fn metadata(&self) -> Metadata { figment::Metadata::named("Pacstall Config") }
+
+    fn data(&self) -> Result<Map<Profile, Dict>, Error> {
+        figment::providers::Serialized::defaults(Config::default()).data()
+    }
+
+    fn profile(&self) -> Option<Profile> { None }
+}
+
+/// The extracted `settings` table.
+#[derive(Deserialize, Debug, Eq, PartialEq, Serialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct Settings {
+    /// The preferred `editor` set by the user. (Defaults to the environment
+    /// variables *EDITOR* -> *VISUAL* -> the editor chosen by the
+    /// `sensible-editor` command.)
+    pub editor: String,
+    /// Number of threads to use for building packages (Defaults to the number
+    /// of *logical cores* on the computer.)
+    pub jobs: u8,
+}
+
+impl Default for Settings {
+    #[allow(clippy::cast_possible_truncation)]
+    fn default() -> Self {
+        Self {
+            editor: env::var("EDITOR").unwrap_or_else(|_| {
+                env::var("VISUAL").unwrap_or_else(|_| "sensible-editor".into())
+            }),
+            jobs: num_cpus::get() as u8,
+        }
+    }
+}
+
+/// The extracted `repositories` array of tables.
+///
+/// Defaults to the official repository.
+#[derive(Deserialize, Debug, Eq, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct Repository {
+    /// The name of the repository.
+    pub name: String,
+    /// URL of the repository.
+    ///
+    /// Note that the URL **isn't verified** during extraction!
+    pub url: String,
+    /// Preference of the repository.
+    ///
+    /// Specifies which repository to look into first during certain operations
+    /// like installing a package. If the package isn't present in the first
+    /// preferred repository, then the second preferred repository is looked
+    /// into.
+    pub preference: u32,
+}
+
+fn default_repository() -> Vec<Repository> { vec![Repository::default()] }
+
+impl Default for Repository {
+    fn default() -> Self {
+        Self {
+            name: "official".into(),
+            url: "https://github.com/pacstall/pacstall-programs".into(),
+            preference: 1,
+        }
+    }
+}
+
+#[cfg(test)]
```

Comment:
Do we include tests in the config.rs or should we make a separate file for testing?

Reply:
All unit tests for the config handler is written in the `tests` module in the `config.rs` file. The unit tests are in the same file as the code.